### DISCUSSION TITLE
How to make a 3-way splitRequestSources() (success, error, ignore)

### DISCUSSION THREAD

**Discussion by colorcube** [state-adapt/state-adapt#59] [Q&A]
This is not necessarily specific to state-apapt.

I want to make a 3-way splitRequestSources() (success, error, ignore) or something similar.

Here's a code excerpt which shows updating an auth token. This is regularly called while the user still might have a valid token.

When the request fails (which can often happen in development) the user gets logged out (loginRefreshRequest.error$). But the user should _only_ logged out with unauthorized/401 error. 
I could catch other errors than 401, but loginRefreshRequest.success$ would be triggered then, without a new token.

I want to ignore all errors except 401.

```typescript
loginRefresh$ = AuthActions.loginRefresh$.pipe(
    concatMap(() =>
        this.authApiService
            .refreshUserFromBackend(this.authApiService.getStorageToken())
            .pipe(toRequestSource('loginRefresh'))
    )
);
loginRefreshRequest = splitRequestSources('loginRefresh', this.loginRefresh$);
    
this.store = adapt(['auth', initialState, adapter], {
    logIn: AuthActions.loginAutoStart$,
    logOut: AuthActions.logout$,
    receiveUser: [this.loginWithTokenRequest.success$, this.loginRefreshRequest.success$],
    receiveError: [
        AuthActions.loginFailure$,
        this.loginWithTokenRequest.error$,
        this.loginRefreshRequest.error$,
    ],
    noop: [this.logoutSuccess$, this.loginAutostarted$],
});
```

**Comment by mfp22** [comment]
Is there a good way to handle this before `this.authApiService.getStorageToken()` returns the observable? I used to provide more options for these utilities but the code was overly complex and I figured people can decide what errors to throw before handing it off to `toRequestSource`. Maybe you could `catchError` and return `NEVER` but re-throw if it's 401. I haven't done this before, tbh.

**Comment by colorcube** [comment]
When this.authApiService.refreshUserFromBackend() returns NEVER from catchError, the stream stops.
The next call of AuthActions.loginRefresh$.next() doesn't trigger the pipe.
I experimented with EMPTY, but still the same.

This works

```typescript
loginRefresh$ = AuthActions.loginRefresh$.pipe(
    concatMap(() =>
        this.authApiService.refreshUserFromBackend(this.authApiService.getStorageToken()).pipe(
            // instead of:
            // toRequestSource('loginRefresh')
            // this:
            toSource('loginRefresh.success$'),
            catchError((error) => {
                if (error.status === 401) {
                    return of(getAction(`loginRefresh.error$`, error));
                } else {
                    return EMPTY;
                }
            }),
        )
    )
);
```

**Reply by mfp22** [reply]
Strange. Yeah, these utilities are proving to be great for the most common use cases, but for any custom behavior they need to be dropped. 

I would just keep what you have then, but maybe create a function similar to `catchErrorSource` but call it `catch401Source` and put your `catchError` code in there. Here's the code for `catchErrorSource` https://github.com/state-adapt/state-adapt/blob/main/libs/rxjs/src/lib/sources/catch-error-source.operator.ts

And as you can see here, `toRequestSource` just wraps `toSource` and `catchErrorSource`. https://state-adapt.github.io/docs/rxjs#torequestsource

I think I need to put links to the source code in the docs for each of these and recommend people look at them and understand they're just for convenience and it's okay to not use them. There are probably 5-10 people using them with no issues for every issue I see opened here, but I'm seeing enough complaints that I think I should do something about it.

**Reply by mfp22** [reply]
Another approach could be to just filter out the non-401 errors in the final `error$` stream produced by `toRequestSource`. 
```typescript
receiveError: this.loginRefreshRequest.error$.pipe(filter(({ payload }) => payload.includes('401'))),
```

**Reply by colorcube** [reply]
I'm fine with my solution, and I think it's just unwrapping toRequestSource. I would say that's ok and no more tools are needed.

The filter is a good idea.
